Android mining interface glitch

## Description
When a mining android breaks an android interface, instead of breaking the block and storing it as an interface, it reset's it back to a standard dispenser.

For example, place an android interface below a miner and use the android miner to break the interface, it will drop a dispenser rather than an interface.

## Server Log
N/A

## Evidence / Proof


## Environment
Slimefun v4.1.11
Minecraft V1.12.1 (We are running ViaVersion on a bungeecord network)
CS-CoreLib V 1.5.15

@Poslovitch did some more testing with debug_fish, and I got an EventException while clicking on it

This is the chat output, as you can see is half way printed:
![2018-01-06_01 10 10](https://user-images.githubusercontent.com/25185835/34633836-17c4677a-f27f-11e7-86e9-8ebd896c54a0.png)


This is the error printed by the console:
`
[01:05:11 ERROR]: Could not pass event PlayerInteractEvent to Slimefun v4.1.12
org.bukkit.event.EventException: null
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.ja                                                                  va:306) ~[spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.jav                                                                  a:62) ~[spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.j                                                                  ava:500) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.j                                                                  ava:485) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPlayerInt                                                                  eractEvent(CraftEventFactory.java:235) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40                                                                  c9ef]
        at net.minecraft.server.v1_12_R1.PlayerInteractManager.a(PlayerInteractM                                                                  anager.java:458) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at net.minecraft.server.v1_12_R1.PlayerConnection.a(PlayerConnection.jav                                                                  a:951) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at net.minecraft.server.v1_12_R1.PacketPlayInUseItem.a(PacketPlayInUseIt                                                                  em.java:37) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at net.minecraft.server.v1_12_R1.PacketPlayInUseItem.a(PacketPlayInUseIt                                                                  em.java:1) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at net.minecraft.server.v1_12_R1.PlayerConnectionUtils$1.run(SourceFile:                                                                  13) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [                                                                  ?:1.8.0_151]
        at java.util.concurrent.FutureTask.run(Unknown Source) [?:1.8.0_151]
        at net.minecraft.server.v1_12_R1.SystemUtils.a(SourceFile:46) [spigot-1.                                                                  12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at net.minecraft.server.v1_12_R1.MinecraftServer.D(MinecraftServer.java:                                                                  748) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at net.minecraft.server.v1_12_R1.DedicatedServer.D(DedicatedServer.java:                                                                  406) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at net.minecraft.server.v1_12_R1.MinecraftServer.C(MinecraftServer.java:                                                                  679) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at net.minecraft.server.v1_12_R1.MinecraftServer.run(MinecraftServer.jav                                                                  a:577) [spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_151]
Caused by: java.lang.NullPointerException
        at me.mrCookieSlime.Slimefun.listeners.ItemListener.debug(ItemListener.j                                                                  ava:121) ~[?:?]
        at sun.reflect.GeneratedMethodAccessor19.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1                                                                  .8.0_151]
        at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_151]
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.ja                                                                  va:302) ~[spigot-1.12.2.jar:git-Spigot-bb614f8-e40c9ef]
        ... 17 more
`

This is miner's inventory (x1 Ancient Pedestal, x1 Android Interface Fuel, x1 Android Interface Items. x1 Block Placer, x1 Grind Stone, x64 Sand):
![2018-01-06_01 16 58](https://user-images.githubusercontent.com/25185835/34633873-60b42010-f27f-11e7-8351-04435fd09733.png)
The block in front of the miner used to be an Electric Furnace, now it's just a furnace with a custom name (no output if clicked with debug_fish).

Still not sure what could be the cause.

My Discord ID is Redemption#2351 by the way.

--- a/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
+++ b/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
@@ -732,11 +732,9 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 		Collection<ItemStack> drops = block.getDrops();
 		if (!blockblacklist.contains(block.getType()) && !drops.isEmpty() && CSCoreLib.getLib().getProtectionManager().canBuild(UUID.fromString(BlockStorage.getBlockInfo(b, "owner")), block)) {
 			try {
-				System.out.println("DEBUG: " + item);
 				if(item != null) {
 					if(fits(b, BlockStorage.retrieve(block))) {
 						pushItems(b, BlockStorage.retrieve(block));
-						System.out.println("DEBUG: IF I'm HERE");
 						if(SlimefunItem.blockhandler.containsKey(item.getID())) SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG);
 						block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
 						
@@ -755,7 +753,6 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 					ItemStack[] items = drops.toArray(new ItemStack[drops.size()]);
 					if (fits(b, items)) {
 						pushItems(b, items);
-						System.out.println("DEBUG: ELSE I'm HERE");
 						block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
 						block.setType(Material.SKULL);
 						block.setData((byte) 1);

--- a/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
+++ b/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
@@ -711,10 +711,14 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 		if (!blockblacklist.contains(block.getType()) && !drops.isEmpty() && CSCoreLib.getLib().getProtectionManager().canBuild(UUID.fromString(BlockStorage.getBlockInfo(b, "owner")), block)) {
 			if(item != null) {
 				if(fits(b, item.getItem())) {
-					pushItems(b, BlockStorage.retrieve(block));
-					if(SlimefunItem.blockhandler.containsKey(item.getID())) SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG);
-					block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
-					block.setType(Material.AIR);
+					if(SlimefunItem.blockhandler.containsKey(item.getID())) {
+						if (SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG)) {
+							pushItems(b, BlockStorage.retrieve(block));
+							if(SlimefunItem.blockhandler.containsKey(item.getID())) SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG);
+							block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
+							block.setType(Material.AIR);
+						}
+					}
 				}
 			}else {
 				ItemStack[] items = drops.toArray(new ItemStack[drops.size()]);
@@ -734,19 +738,23 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 			try {
 				if(item != null) {
 					if(fits(b, item.getItem())) {
-						pushItems(b, BlockStorage.retrieve(block));
-						if(SlimefunItem.blockhandler.containsKey(item.getID())) SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG);
-						block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
-						
-						block.setType(Material.SKULL);
-						block.setData((byte) 1);
-	
-						Skull skull = (Skull) block.getState();
-						skull.setRotation(face);
-						skull.update(true, false);
-						CustomSkull.setSkull(block, CustomSkull.getTexture(getItem()));
-						b.setType(Material.AIR);
-						BlockStorage.moveBlockInfo(b, block);
+						if(SlimefunItem.blockhandler.containsKey(item.getID())) {
+							if (SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG)) {
+								pushItems(b, BlockStorage.retrieve(block));
+								
+								block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
+								
+								block.setType(Material.SKULL);
+								block.setData((byte) 1);
+			
+								Skull skull = (Skull) block.getState();
+								skull.setRotation(face);
+								skull.update(true, false);
+								CustomSkull.setSkull(block, CustomSkull.getTexture(getItem()));
+								b.setType(Material.AIR);
+								BlockStorage.moveBlockInfo(b, block);
+							}
+						}
 					}
 				}
 				else {

--- a/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
+++ b/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
@@ -316,20 +316,17 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 					}
 					case DIG_FORWARD: {
 						Block block = b.getRelative(BlockFace.valueOf(BlockStorage.getBlockInfo(b, "rotation")));
-						SlimefunItem item = BlockStorage.check(block);
-						mine(b, block, item);
+						mine(b, block);
 						break;
 					}
 					case DIG_UP: {
 						Block block = b.getRelative(BlockFace.UP);
-						SlimefunItem item = BlockStorage.check(block);
-						mine(b, block, item);
+						mine(b, block);
 						break;
 					}
 					case DIG_DOWN: {
 						Block block = b.getRelative(BlockFace.DOWN);
-						SlimefunItem item = BlockStorage.check(block);
-						mine(b, block, item);
+						mine(b, block);
 						break;
 					}
 					case CATCH_FISH: {
@@ -347,22 +344,19 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 					case MOVE_AND_DIG_FORWARD: {
 						BlockFace face = BlockFace.valueOf(BlockStorage.getBlockInfo(b, "rotation"));
 						Block block = b.getRelative(face);
-						SlimefunItem item = BlockStorage.check(block);
-						movedig(b, face, block, item);
+						movedig(b, face, block);
 						break;
 					}
 					case MOVE_AND_DIG_UP: {
 						BlockFace face = BlockFace.valueOf(BlockStorage.getBlockInfo(b, "rotation"));
 						Block block = b.getRelative(BlockFace.UP);
-						SlimefunItem item = BlockStorage.check(block);
-						movedig(b, face, block, item);
+						movedig(b, face, block);
 						break;
 					}
 					case MOVE_AND_DIG_DOWN: {
 						BlockFace face = BlockFace.valueOf(BlockStorage.getBlockInfo(b, "rotation"));
 						Block block = b.getRelative(BlockFace.DOWN);
-						SlimefunItem item = BlockStorage.check(block);
-						movedig(b, face, block, item);
+						movedig(b, face, block);
 						break;
 					}
 					case INTERFACE_ITEMS: {
@@ -706,9 +700,10 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 		}
 	}
 
-	private void mine(Block b, Block block, SlimefunItem item) {
+	private void mine(Block b, Block block) {
 		Collection<ItemStack> drops = block.getDrops();
 		if (!blockblacklist.contains(block.getType()) && !drops.isEmpty() && CSCoreLib.getLib().getProtectionManager().canBuild(UUID.fromString(BlockStorage.getBlockInfo(b, "owner")), block)) {
+			SlimefunItem item = BlockStorage.check(block);
 			if(item != null) {
 				if(fits(b, item.getItem())) {
 					if(SlimefunItem.blockhandler.containsKey(item.getID())) {
@@ -732,10 +727,11 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 	}
 
 	@SuppressWarnings("deprecation")
-	private void movedig(Block b, BlockFace face, Block block, SlimefunItem item) {
+	private void movedig(Block b, BlockFace face, Block block) {
 		Collection<ItemStack> drops = block.getDrops();
 		if (!blockblacklist.contains(block.getType()) && !drops.isEmpty() && CSCoreLib.getLib().getProtectionManager().canBuild(UUID.fromString(BlockStorage.getBlockInfo(b, "owner")), block)) {
 			try {
+				SlimefunItem item = BlockStorage.check(block);
 				if(item != null) {
 					if(fits(b, item.getItem())) {
 						if(SlimefunItem.blockhandler.containsKey(item.getID())) {

--- a/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
+++ b/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
@@ -710,7 +710,7 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 		Collection<ItemStack> drops = block.getDrops();
 		if (!blockblacklist.contains(block.getType()) && !drops.isEmpty() && CSCoreLib.getLib().getProtectionManager().canBuild(UUID.fromString(BlockStorage.getBlockInfo(b, "owner")), block)) {
 			if(item != null) {
-				if(fits(b, BlockStorage.retrieve(block))) {
+				if(fits(b, item.getItem())) {
 					pushItems(b, BlockStorage.retrieve(block));
 					if(SlimefunItem.blockhandler.containsKey(item.getID())) SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG);
 					block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
@@ -733,7 +733,7 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 		if (!blockblacklist.contains(block.getType()) && !drops.isEmpty() && CSCoreLib.getLib().getProtectionManager().canBuild(UUID.fromString(BlockStorage.getBlockInfo(b, "owner")), block)) {
 			try {
 				if(item != null) {
-					if(fits(b, BlockStorage.retrieve(block))) {
+					if(fits(b, item.getItem())) {
 						pushItems(b, BlockStorage.retrieve(block));
 						if(SlimefunItem.blockhandler.containsKey(item.getID())) SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG);
 						block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());

--- a/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
+++ b/src/me/mrCookieSlime/Slimefun/Android/ProgrammableAndroid.java
@@ -316,17 +316,20 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 					}
 					case DIG_FORWARD: {
 						Block block = b.getRelative(BlockFace.valueOf(BlockStorage.getBlockInfo(b, "rotation")));
-						mine(b, block);
+						SlimefunItem item = BlockStorage.check(block);
+						mine(b, block, item);
 						break;
 					}
 					case DIG_UP: {
 						Block block = b.getRelative(BlockFace.UP);
-						mine(b, block);
+						SlimefunItem item = BlockStorage.check(block);
+						mine(b, block, item);
 						break;
 					}
 					case DIG_DOWN: {
 						Block block = b.getRelative(BlockFace.DOWN);
-						mine(b, block);
+						SlimefunItem item = BlockStorage.check(block);
+						mine(b, block, item);
 						break;
 					}
 					case CATCH_FISH: {
@@ -344,19 +347,22 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 					case MOVE_AND_DIG_FORWARD: {
 						BlockFace face = BlockFace.valueOf(BlockStorage.getBlockInfo(b, "rotation"));
 						Block block = b.getRelative(face);
-						movedig(b, face, block);
+						SlimefunItem item = BlockStorage.check(block);
+						movedig(b, face, block, item);
 						break;
 					}
 					case MOVE_AND_DIG_UP: {
 						BlockFace face = BlockFace.valueOf(BlockStorage.getBlockInfo(b, "rotation"));
 						Block block = b.getRelative(BlockFace.UP);
-						movedig(b, face, block);
+						SlimefunItem item = BlockStorage.check(block);
+						movedig(b, face, block, item);
 						break;
 					}
 					case MOVE_AND_DIG_DOWN: {
 						BlockFace face = BlockFace.valueOf(BlockStorage.getBlockInfo(b, "rotation"));
 						Block block = b.getRelative(BlockFace.DOWN);
-						movedig(b, face, block);
+						SlimefunItem item = BlockStorage.check(block);
+						movedig(b, face, block, item);
 						break;
 					}
 					case INTERFACE_ITEMS: {
@@ -700,36 +706,67 @@ public abstract class ProgrammableAndroid extends SlimefunItem {
 		}
 	}
 
-	private void mine(Block b, Block block) {
+	private void mine(Block b, Block block, SlimefunItem item) {
 		Collection<ItemStack> drops = block.getDrops();
 		if (!blockblacklist.contains(block.getType()) && !drops.isEmpty() && CSCoreLib.getLib().getProtectionManager().canBuild(UUID.fromString(BlockStorage.getBlockInfo(b, "owner")), block)) {
-			ItemStack[] items = drops.toArray(new ItemStack[drops.size()]);
-			if (fits(b, items)) {
-				pushItems(b, items);
-				block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
-				block.setType(Material.AIR);
+			if(item != null) {
+				if(fits(b, BlockStorage.retrieve(block))) {
+					pushItems(b, BlockStorage.retrieve(block));
+					if(SlimefunItem.blockhandler.containsKey(item.getID())) SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG);
+					block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
+					block.setType(Material.AIR);
+				}
+			}else {
+				ItemStack[] items = drops.toArray(new ItemStack[drops.size()]);
+				if (fits(b, items)) {
+					pushItems(b, items);
+					block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
+					block.setType(Material.AIR);
+				}
 			}
 		}
 	}
 
 	@SuppressWarnings("deprecation")
-	private void movedig(Block b, BlockFace face, Block block) {
+	private void movedig(Block b, BlockFace face, Block block, SlimefunItem item) {
 		Collection<ItemStack> drops = block.getDrops();
 		if (!blockblacklist.contains(block.getType()) && !drops.isEmpty() && CSCoreLib.getLib().getProtectionManager().canBuild(UUID.fromString(BlockStorage.getBlockInfo(b, "owner")), block)) {
 			try {
-				ItemStack[] items = drops.toArray(new ItemStack[drops.size()]);
-				if (fits(b, items)) {
-					pushItems(b, items);
-					block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
-					block.setType(Material.SKULL);
-					block.setData((byte) 1);
-
-					Skull skull = (Skull) block.getState();
-					skull.setRotation(face);
-					skull.update(true, false);
-					CustomSkull.setSkull(block, CustomSkull.getTexture(getItem()));
-					b.setType(Material.AIR);
-					BlockStorage.moveBlockInfo(b, block);
+				System.out.println("DEBUG: " + item);
+				if(item != null) {
+					if(fits(b, BlockStorage.retrieve(block))) {
+						pushItems(b, BlockStorage.retrieve(block));
+						System.out.println("DEBUG: IF I'm HERE");
+						if(SlimefunItem.blockhandler.containsKey(item.getID())) SlimefunItem.blockhandler.get(item.getID()).onBreak(null, block, item, UnregisterReason.ANDROID_DIG);
+						block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
+						
+						block.setType(Material.SKULL);
+						block.setData((byte) 1);
+	
+						Skull skull = (Skull) block.getState();
+						skull.setRotation(face);
+						skull.update(true, false);
+						CustomSkull.setSkull(block, CustomSkull.getTexture(getItem()));
+						b.setType(Material.AIR);
+						BlockStorage.moveBlockInfo(b, block);
+					}
+				}
+				else {
+					ItemStack[] items = drops.toArray(new ItemStack[drops.size()]);
+					if (fits(b, items)) {
+						pushItems(b, items);
+						System.out.println("DEBUG: ELSE I'm HERE");
+						block.getWorld().playEffect(block.getLocation(), Effect.STEP_SOUND, block.getType());
+						block.setType(Material.SKULL);
+						block.setData((byte) 1);
+	
+						Skull skull = (Skull) block.getState();
+						skull.setRotation(face);
+						skull.update(true, false);
+						CustomSkull.setSkull(block, CustomSkull.getTexture(getItem()));
+						b.setType(Material.AIR);
+						BlockStorage.moveBlockInfo(b, block);
+					}
 				}
 			} catch(Exception x) {
 				x.printStackTrace();

--- a/src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/UnregisterReason.java
+++ b/src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/UnregisterReason.java
@@ -17,6 +17,11 @@ public enum UnregisterReason {
 	/**
 	 * A player breaks the block.
 	 */
-	PLAYER_BREAK
+	PLAYER_BREAK,
+	
+	/**
+	 * An android miner breaks the block.
+	 */
+	ANDROID_DIG
 	
 }

--- a/src/me/mrCookieSlime/Slimefun/api/BlockStorage.java
+++ b/src/me/mrCookieSlime/Slimefun/api/BlockStorage.java
@@ -246,6 +246,15 @@ public class BlockStorage {
 		addBlockInfo(block, "id", item, true);
 	}
 	
+	/**
+     * Retrieves the SlimefunItem's ItemStack from the specified Block.
+     * If the specified Block is registered in BlockStorage, its data will be erased from it, regardless of the returned value.
+     *
+     * @param  block  the block to retrieve the ItemStack from
+     * @return the SlimefunItem's ItemStack corresponding to the block if it has one, otherwise null
+     *
+     * @since 4.0
+     */
 	public static ItemStack retrieve(Block block) {
 		if (!hasBlockInfo(block)) return null;
 		else {

